substratevm/src/com.oracle.svm.driver/src/com/oracle/svm/driver/MemoryUtil.java

```diff
@@ -82,23 +93,40 @@ public static List<String> determineMemoryFlags(NativeImage.HostFlags hostFlags)
     }
 
     /**
-     * Returns a percentage (0.0-100.0) to be used as a value for the -XX:MaxRAMPercentage flag of
-     * the builder process. Prefer free memory over total memory to reduce memory pressure on the
-     * host machine. Note that this method uses OperatingSystemMXBean, which is container-aware.
+     * Returns a percentage (0.0-100.0) to be used as a value for the -XX:MaxRAMPercentage or
+     * -XX:MaximumHeapSizePercent flags of the builder process. Dedicated mode uses a fixed
+     * percentage of total memory and is the default in containers. Shared mode tries to use
+     * available memory to reduce memory pressure on the host machine. Note that this method uses
+     * OperatingSystemMXBean, which is container-aware.
      */
-    private static double determineReasonableMaxRAMPercentage() {
+    private static List<String> determineReasonableMaxRAMPercentage(Function<Double, String> toMemoryFlag) {
         var osBean = (com.sun.management.OperatingSystemMXBean) ManagementFactory.getOperatingSystemMXBean();
         final double totalMemorySize = osBean.getTotalMemorySize();
-        double reasonableMaxMemorySize = osBean.getFreeMemorySize();
+        final double dedicatedMemorySize = totalMemorySize * DEDICATED_MODE_TOTAL_MEMORY_RATIO;
 
-        if (reasonableMaxMemorySize < DEDICATED_MODE_THRESHOLD) {
-            /*
-             * When free memory is low, for example in memory-constrained environments or when a
-             * good amount of memory is used for caching, use a fixed percentage of total memory
-             * rather than free memory. In containerized environments, builds are expected to run
-             * more or less exclusively (builder + driver + optional Gradle/Maven process).
-             */
-            reasonableMaxMemorySize = totalMemorySize * DEDICATED_MODE_TOTAL_MEMORY_RATIO;
+        String memoryUsageReason = "unknown";
+        final boolean isDedicatedMemoryUsage;
+        if ("true".equals(System.getenv("CI"))) {
```

Comment:
Shouldn't this be?

```suggestion
        if (SubstrateUtil.isRunningInCI) {
```

Reply:
Good point, although isRunningInCI() also uses `isTTY()`.